Add file system change notifier

davidbrochart opened the pull request:
I think we should adhere to [watchfiles](https://watchfiles.helpmanual.io)'s API.

Closes #778.

blink1073 commented:
I verified that watchfiles is available on conda-forge and has an MIT license.

davidbrochart commented:
> I verified that watchfiles is available on conda-forge

I packaged it yesterday :smile:

fcollonval commented:
I was about to ask how it compares with watchgod - but actually that is the new name :smile:

kevin-bates commented:
> I was about to ask how it compares with watchgod - but actually that is the new name

We use [`watchdog`](https://pypi.org/project/watchdog/) in EG and Elyra for cache consistency implementations.  I'm guessing `watchgod `(`watchfiles`) came afterward (based on the play on the name), and I'm curious if there's something that should be avoided about `watchdog`?  Since both EG and Elyra depend on jupyter_server, I'd rather not require our users to install two packages that essentially do the same thing.  I'm fine converting the `watchdog` instances, but would like to better understand why `watchfiles` is advantageous.

> I verified that watchfiles is available on conda-forge and has an MIT license.

Hmm.  I see `watchdog` uses an Apache license.  Is this difference a licensing thing?  If so, are there also functionality differences that warrant the use of `watchfiles` over `watchdog`?

blink1073 commented:
Apache is fine, I was mainly checking for not GPL

davidbrochart commented:
@kevin-bates I had a previous experience with `watchdog`, my conclusion was that it is much better to work with `watchgod` (and thus `watchfiles`) in an async context. Also, `watchfiles` has some nice features such as grouping events, I'm not sure if `watchdog` has the equivalent.

davidbrochart commented:
The question here is: if we are importing `watchfiles` (in the Python meaning), does it mean we are also importing its API as a specification that other contents manager should be compliant with?

davidbrochart commented:
In JupyterLab I want to [check if the contents manager supports watching file changes](https://github.com/jupyterlab/jupyterlab/blob/76f8c151a5c50162a9743439424dceae7441fd21/jupyterlab/handlers/yjs_echo_ws.py#L99). For that I need to know if the underlying file system is the locacl file system:
```py
if isinstance(self.contents_manager, (FileManagerMixin, AsyncFileManagerMixin)):
```
But this is not ideal since custom contents managers could target the local file system and not derive from `FileManagerMixin` or `AsyncFileManagerMixin`.
I think the [ContentsManager class](https://github.com/jupyter-server/jupyter_server/blob/64bfcea7e802216d92f1947120a4e2c8bcf5b286/jupyter_server/services/contents/manager.py#L37) should have a `watchfiles` property returning `None` by default, that implementations should override if they can offer this service. For instance, our file managers would have:
```py
@property
def watchfiles(self):
    import watchfiles
    return watchfiles.__all__
    # ('watch', 'awatch', 'run_process', 'arun_process', 'Change', 'BaseFilter', 'DefaultFilter', 'PythonFilter', 'VERSION')
```
Then, the user could inspect the objects returned by this property to know what is supported.
This makes the file system change notifyer a "loose specification" in the contents manager, meaning that it tries to follow [watchfiles](https://watchfiles.helpmanual.io)' API when possible.
An alternative solution would be to fully specify the API in the contents manager, maybe by copying a subset of watchfiles' API, but it seems less flexible. What do you think?

davidbrochart commented:
In this last commit I added the `watchfiles` attribute in the contents manager class. In `ContentsManager` it is implemented as an instance of an empty class, and in `FileContentsManager` and `AsyncFileContentsManager` it is the `watchfiles` module.

martinRenou commented:
> In this last commit I added the watchfiles attribute in the contents manager class

I am wondering what does that mean in terms of jupyter-server semantic versioning. I guess it means jupyter-server will have to follow watchfiles if it adopts its API? I suppose we don't want to be in a situation where watchfile don't follow semver and breaks on a patch release?

If the watchfiles API is not too complicated, could the ContentManager just implement the methods it provides and privately make use of watchfiles?

davidbrochart commented:
Yes that's what I meant with "loose specification". On the other hand, even if we implement the methods in ContentsManager and under the hood delegate to watchfiles, we still have no control over changes e.g. in the return values, etc. Or we would have to exactly pin the package, and manually sync with new releases. But maybe it is safer?

trungleduc commented on jupyter_server/services/contents/manager.py:
```diff
@@ -34,6 +34,13 @@
 copy_pat = re.compile(r"\-Copy\d*\.")
 
 
+class NoWatchfilesAPI:
```
I think `NoWatchfilesAPI` should contain the desired APIs without concrete implementations. So it can be used as a type hint for the `watchfiles` property of `ContentsManager`

davidbrochart replied:
Yes so that would go in [Martin's direction](https://github.com/jupyter-server/jupyter_server/pull/783#issuecomment-1100117524).

kevin-bates commented:
Thanks for working on this @davidbrochart.  I agree with @martinRenou and believe we should introduce some form of abstraction that allows folks the ability to implement their own "watchfiles" integration (in their own `ContentsManager`).  We should also not preclude custom contents managers from wanting to watch _remote_ files either - so filesystem locality should not be relevant - and boils down to _does the configured `ContentsManager` support file-change notification or not?_

davidbrochart commented:
@kevin-bates @martinRenou @trungleduc I added a `BaseWatchfiles` class, that is currently populated with the following attributes (links to the corresponding watchfiles API):
- [watch](https://watchfiles.helpmanual.io/api/watch/#watchfiles.watch)
- [awatch](https://watchfiles.helpmanual.io/api/watch/#watchfiles.awatch)
- [Change](https://watchfiles.helpmanual.io/api/watch/#watchfiles.Change)

I believe this is the minimum necessary for now (at least for my use-case :smile: ). We can add more in the future if needed. I also pinned `watchfiles==0.13`, so that this is a stable API in our contents manager implementations.

kevin-bates commented:
Isn't this just exposing the `watchfiles` API?  I think we need to ask what we need as a notification item and expose _that_.  This is requiring that all implementations implement `watchfiles`.  

What do we need in addition to the file and its corresponding action: added, modified, or deleted? 

Are we going to allow subsets of `src_root` to be monitored or is it all or nothing under `src_root`?  Perhaps we could have a configurable _list_ that specifies the directories to be monitored?  And can the kinds of files be specified - `*.ipynb`, `*.py`, etc. relative to each monitored directory?  If so, we could _include_ the filtering expression in the directory:

```python
c.ContentsManager.watched_directories = ["foo/bar/{*.ipynb, *.py}", "baz/{*.yaml}"]
```

or, if only `src_root`, just let the configurable be the filter...
```python
c.ContentsManager.watched_files = ["*.ipynb", "*.py", "*.yaml"]
```
and non-None values indicate that "watch files" is enabled.

davidbrochart commented:
> Isn't this just exposing the `watchfiles` API? I think we need to ask what we need as a notification item and expose _that_. This is requiring that all implementations implement `watchfiles`.

I think we should adhere to `watchfiles`' API, but that doesn't mean that all implementations must implement it. Our file contents managers implement all of it, and directly delegates to `watchfiles`. But other implementations can implement a subset of it. But what they do implement, should conform to the API.

> What do we need in addition to the file and its corresponding action: added, modified, or deleted?
> 
> Are we going to allow subsets of `src_root` to be monitored or is it all or nothing under `src_root`? Perhaps we could have a configurable _list_ that specifies the directories to be monitored? And can the kinds of files be specified - `*.ipynb`, `*.py`, etc. relative to each monitored directory? If so, we could _include_ the filtering expression in the directory:
> 
> ```python
> c.ContentsManager.watched_directories = ["foo/bar/{*.ipynb, *.py}", "baz/{*.yaml}"]
> ```
> 
> or, if only `src_root`, just let the configurable be the filter...
> 
> ```python
> c.ContentsManager.watched_files = ["*.ipynb", "*.py", "*.yaml"]
> ```
> 
> and non-None values indicate that "watch files" is enabled.

I think all your suggestions can be implemented with the `watch_filter` argument of `watch()`.
That is why I think we should adhere to `watchfiles`' API, there is no need to reinvent the wheel. It is a very well designed library and I doubt we would come up with a better API. And anyway, if we want to use it under the hood at least for our file contents managers, having another layer on top of its original API will probably make things more complicated, don't you think?

kevin-bates commented:
The idea would be to _constrain_ what is exposed via the API, otherwise, folks that don't use `watchfiles` within their implementation will need to figure out how to implement aspects of `watchfiles` that don't exist in other offerings because our users will have grown accustomed to and, worse, _expect_ the `watchfiles` functionality if its exposed in the raw.  This is why I asked what we _minimally need_ and only expose that as the "API".  We can use whatever functionality we want internally.

davidbrochart commented:
I get your point, on the other hand it would be nice to offer various level of functionalities depending on the file system. On a local file system, I would expect to be able to use all of `watchfiles`' functionalities, and on a less capable file system, less functionalities. Maybe we're only talking about being able to watch file changes vs *not* being able to (and actually I need that in JupyterLab), maybe it's more subtle, like supporting the `debounce` parameter of `watch()` (which affects how changes are grouped).
That raises the question of letting the user discover what functionalities are supported. If no functionality at all, calling `watch()` currently raises an exception, so that's one way. For more subtle functionalities, I was thinking that the user could inspect the signature of `watch()`, and see e.g. if `debounce` is present, but I agree that it's not ideal. Maybe we could have a method that would return the list of supported arguments?
Or maybe, as you suggested, we should support a fixed set of functionalities, that would be the minimum that could be supported by all file systems. Thanks for you feedback Kevin!

Zsailer commented:
We spent some time triaging PRs in the Server Meeting today. What's the current status of this PR? 

It seems to me that it might be better to offer this as an alternative implementation of the FileContentsManager that could be provided as a separate package/extension. Imposing a "file-watching" cost on all users can be quite expensive, so making it an opt-in feature might be better.

davidbrochart commented:
TO;DR
(too old; don't remember :smile:)
